SignalServiceKit/src/Util/OWSBackgroundTask.m

```diff
@@ -4,14 +4,290 @@
+            [CurrentAppContext() endBackgroundTask:backgroundTaskId];
+            return YES;
+        }
+    }
+}
+
+// Returns NO if the background task cannot be begun.
+- (BOOL)startBackgroundTask
+{
+    OWSAssert(CurrentAppContext().isMainApp);
+
+    @synchronized(self)
+    {
+        OWSAssert(self.backgroundTaskId == UIBackgroundTaskInvalid);
+
+        self.backgroundTaskId = [CurrentAppContext() beginBackgroundTaskWithExpirationHandler:^{
+            // Supposedly [UIApplication beginBackgroundTaskWithExpirationHandler]'s handler
+            // will always be called on the main thread, but in practice we've observed
+            // otherwise.
+            //
+            // See:
+            // https://developer.apple.com/documentation/uikit/uiapplication/1623031-beginbackgroundtaskwithexpiratio)
+            OWSAssert([NSThread isMainThread]);
+
+            [self backgroundTaskExpired];
+        }];
+
+        // If a background task could not be begun, call the completion block.
+        if (self.backgroundTaskId == UIBackgroundTaskInvalid) {
+            DDLogError(@"%@ background task could not be started.", self.logTag);
+
+            return NO;
+        }
+        return YES;
+    }
+}
+
+- (void)backgroundTaskExpired
+{
+    UIBackgroundTaskIdentifier backgroundTaskId;
+    NSDictionary<OWSTaskId, BackgroundTaskExpirationBlock> *expirationMap;
+
+    @synchronized(self)
+    {
+        backgroundTaskId = self.backgroundTaskId;
+        self.backgroundTaskId = UIBackgroundTaskInvalid;
+
+        expirationMap = [self.expirationMap copy];
+        [self.expirationMap removeAllObjects];
+    }
+
+    // It'd be nice to do this work synchronously, but it seems unsafe to
+    // depend on all expiration blocks being cheap.
+    //
+    // OWSBackgroundTask's API guarantees that completionBlock will always
+    // be called on the main thread, so facilitate that by dispatching async
+    // to main queue here.  That way we can ensure that we don't end the
+    // background task until all of the completion blocks have completed.
+    dispatch_async(dispatch_get_main_queue(), ^{
+        for (BackgroundTaskExpirationBlock expirationBlock in expirationMap) {
+            expirationBlock();
+        }
+        if (backgroundTaskId != UIBackgroundTaskInvalid) {
+            // Apparently we need to "end" even expired background tasks.
+            [CurrentAppContext() endBackgroundTask:backgroundTaskId];
+        }
+    });
+}
+
+- (void)timerDidFire
+{
+    @synchronized(self)
+    {
+        [self.continuityTimer invalidate];
+        self.continuityTimer = nil;
+
+        [self ensureBackgroundTaskState];
+    }
+}
+
+@end
+
+#pragma mark -
+
 @interface OWSBackgroundTask ()
 
 @property (nonatomic, readonly) NSString *label;
 
 // This property should only be accessed while synchronized on this instance.
-@property (nonatomic) UIBackgroundTaskIdentifier backgroundTaskId;
+@property (nonatomic, nullable) OWSTaskId taskId;
```

Comment:
There's a new implementation detail in this file - "task ids" which are a token used to identify instances of `OWSBackgroundTask` for the manager.